feat: test bento bundle

asafalinadsg opened the pull request:
## What does this PR address?
<!--
Thanks for sending a pull request!

Congrats for making it this far! Here's a 🍱 for you. There are still a few steps ahead.

Please make sure to read the contribution guidelines, then fill out the blanks below before requesting a code review.

Name your Pull Request with one of the following prefixes, e.g. "feat: add support for PyTorch", to indicate the type of changes proposed. This is based on the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary).
  - feat: (new feature for the user, not a new feature for build script)
  - fix: (bug fix for the user, not a fix to a build script)
  - docs: (changes to the documentation)
  - style: (formatting, missing semicolons, etc; no production code change)
  - refactor: (refactoring production code, eg. renaming a variable)
  - perf: (code changes that improve performance)
  - test: (adding missing tests, refactoring tests; no production code change)
  - chore: (updating grunt tasks etc; no production code change)
  - build: (changes that affect the build system or external dependencies)
  - ci: (changes to configuration files and scripts)
  - revert: (reverts a previous commit)

Describe your changes in detail. Attach screenshots here if appropriate.

Once you're done with this, someone from BentoML team or community member will help review your PR (see "Who can help review?" section for potential reviewers.). If no one has reviewed your PR after a week have passed, don't hesitate to post a new comment and ping @-the same person. Notifications sometimes get lost 🥲.
-->
This PR adds the `bentoml test` command which enables testing a Bento for making sure it is correctly packaged.
In order to use `bentoml test` the Bento builder must supply input/output example(s), this will be done in the `bentofile.yaml` in a dedicated section called `endpoints_tests`. 
Once specified, the Bento builder can call `bentoml test <bento-tag>` to test the specified bento, or alternatively call `bentoml build --test` which will build and test in one call (still not implemented).

The testing itself will invoke the `bentoml containerize` command to create an image, will launch a docker container from that image and then will interact with the docker through HTTP request to test the endpoints. This will cover all the critical functionality of Bentos; creating an environment from the Bento specification and the service's endpoints.


## Example

Imagine I have the below, dummy, scikit-learn scaler model serving:
```python
# init
model_name = 'test-scaler'

# load model's metadata
scaler_meta = bentoml.sklearn.get(f'{model_name}:latest')

# create service
scaler_runner = scaler_meta.to_runner()
scaler_service = bentoml.Service(model_name, runners=[scaler_runner])

# create endpoint for the service
@scaler_service.api(input=PandasDataFrame(), output=NumpyNdarray())
def my_predict_endpoint(input_df: pd.DataFrame) -> pd.DataFrame:
    res = scaler_runner.run(input_df)
    return res
```

The altered `Bentofile.yaml` will look as follows (note the new `endpoints_tests` section):
```yaml
service: "serve_model:scaler_service"  # Same as the argument passed to `bentoml serve`
labels:
  owner: bentoml-team
  stage: dev
include:
  - "*.py"  # A pattern for matching which files to include in the bento
python:
  packages: # Additional pip packages required by the service
    - scikit-learn
    - pandas

## TESTING
endpoints_tests:
  my_predict_endpoint:
    - input: [[1, 2, 3, 4, 5]]
      output: [[-1.4638501094227998,-1.4638501094227998,-0.8783100656536799,-0.29277002188455997,-0.8783100656536799]]
    - input: [[1, 2, 3, 4, 5]]
```

As you can see, I made 2 tests for `my_predict_endpoint` endpoint:
1. one that takes both input and output.
2. one that takes only an input.

Now, when building a Bento, using `bentoml build`, we will get a Bento tag, somewhat similar to `test-scaler:e45pn7bm7cr4rdsd`.


Now we can test our newly created Bento, using the command `bentoml test test-scaler:e45pn7bm7cr4rdsd`, which will do the following:
1. Create a docker image from the given Bento (using the `containerize` command under the hood).
2. Run a docker container from the newly created docker image.
3. Run the 2 tests, specified in the `Bentofile.yaml`, against the `my_predict_endpoint` endpoint:
i. The 1st test, which takes the input, send it to the endpoint, and check if the received result equals to the expected output specified in the test (`[[-1.4638501094227998, ...]]`).
ii. The 2nd test, which takes the input, send it to the endpoint, and just check that the service didn't throw and error (unlike the 1st test, it doesnt do any checking on the responded output). 


<!-- Remove if not applicable -->
## Issue
https://github.com/bentoml/BentoML/issues/2967

## Before submitting:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- If you plan to update documentation or tests in follow-up, please note -->
- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

## Who can help review?
@aarnphm
@parano

Feel free to tag members/contributors who can help review your PR.
<!--
Feel free to ping any of the BentoML members for help on your issue, but don't ping more than three people 😊.
If you know how to use git blame, that is probably the easiest way.


Team members that you can ping:
- @parano
- @yubozhao
- @bojiang
- @ssheng
- @aarnphm
- @sauyon
- @larme
- @yetone
- @jjmachan
-->

pep8speaks commented:
Hello @asafalinadsg! Thanks for opening this PR.











* In the file [`bentoml_cli/test.py`](https://github.com/bentoml/BentoML/blob/32896a37ac21dc0097507879f05784f022982c90/bentoml_cli/test.py):

> [Line 3:1](https://github.com/bentoml/BentoML/blob/32896a37ac21dc0097507879f05784f022982c90/bentoml_cli/test.py#L3): [F401](https://duckduckgo.com/?q=pep8%20F401) 'sys' imported but unused
> [Line 4:1](https://github.com/bentoml/BentoML/blob/32896a37ac21dc0097507879f05784f022982c90/bentoml_cli/test.py#L4): [F401](https://duckduckgo.com/?q=pep8%20F401) 'typing as t' imported but unused

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

yubozhao commented:
@asafalinadsg this amazing

@ssheng Can we check this one out?

aarnphm commented:
Hi there, one small request. Is it possible that you can move the `client` folder to `bentoml`?

I think it is nice to also expose the python api
```python
from bentoml.client import testclient
...
```

aarnphm reviewed (commented):
Some preliminary comments with implementation details.

Generally our strategy with typing is we adhere to PEP604, and if anything related to types that doesn't affect runtime, then it is safe to put it under TYPE_CHECKING block.

I think the ability to include testing data under the `bentofile.yaml` is nice and convenient, but I'm not convinced yet if this is the way going forward.

I might have to think more about the design of this, but absolutely love the initial work 😄

asafalinadsg commented:
I've added few more things.
The `bentofile.yaml` file changed a bit:
Notice the input / output option of local / remote (s3 etc...) files.

```yaml
service: "serve_model:scaler_service"  # Same as the argument passed to `bentoml serve`
labels:
  owner: bentoml-team
  stage: dev
include:
  - "*.py"  # A pattern for matching which files to include in the bento
python:
  packages: # Additional pip packages required by the service
    - scikit-learn
    - pandas

# TESTING
tests:
  config:
    timeout: 10
  endpoints:
    my_predict_endpoint:
      - input: [[1, 2, 3, 4, 5]]
        output: [[1 ,2, 3, 4, 5]]
      - input: [[1, 2, 3, 4, 5]]
      - input: /my/local/file/path/input.json
        output: /my/remote/file/path/output.json

```

Also, I've added the option to `bentoml build --test`.
This will create a Bento, as usual, plus run the tests right after the build.
In other words:
```bash
bentoml build
bentoml test <my-bento>
``` 
is just like:
```bash
bentoml build --test
```

aarnphm commented:
> > Notice the input / output option of local / remote (s3 etc...) files.

We are currently using PyFilesystem2, would be great if we can use that.

asafalinadsg commented on bentoml/client/test_client/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from bentoml.client.test_client.utils import is_equal
+from bentoml.client.test_client.utils import get_test_data
+from bentoml.client.test_client.client import Endpoint
+from bentoml.client.test_client.client import TestClient
```
cool thing, didnt know of the `__all__` 👍 
thanks :)

asafalinadsg commented on bentoml/bentos.py:
```diff
@@ -477,6 +477,118 @@ def containerize(
+    )
+
+    # setup bento client
+    bento_client = TestClient()
```
Interesting...
thanks @aarnphm 
I think it's better to wait with this one, cause to my understanding there should be a bento client support coming soon,
and therefore maybe it will be better using it instead of the `TestClient` I created.
So this one is to be continued...

asafalinadsg commented on bentoml/client/test_client/utils.py:
```diff
@@ -0,0 +1,57 @@
+import typing as t
```
@aarnphm , I wonder:
I'm using python3.9 locally.
PEP604 support is from python 3.10 onwards.
I wonder, how can I make it work on python3.9?

Was trying to replace this
```python
port: t.Union[int, str] = 3000
```
with this:
```python
port: int | str = 3000,
```
But when running I'm getting the error:
`TypeError: unsupported operand type(s) for |: 'type' and 'type'`

What am I missing? I'm sure I am missing something...

(note it has nothing to do with `bentoml/client/test_client/utils.py`)

aarnphm replied:
PEP604 requires you to also add `from __future__ import annotations`

asafalinadsg commented:
@aarnphm 

> I think the ability to include testing data under the bentofile.yaml is nice and convenient, but I'm not convinced yet if this is the way going forward.

Do you still consider adding testing as part of the `bentofile.yaml`?
Are there any other alternative ways you consider for testing a bento?

ssheng commented:
@asafalinadsg Sorry about the delay response. The team discussed internally about the proposal and we liked the idea of make bentos testable. We also recognize the challenges of building bentos without being able to test it in the environment (e.g. docker) that it is going to be run in. We propose a 2-step change to make the `bentoml test` feature more useful.

1. Making bentos testable through `bentoml test`. As you implemented in this PR, we can add a `tests` section in the `bentofile.yaml` for the bento creator to define the tests to be run against the deployed bento. However, instead of hard coding inputs and outputs in the `bentofile.yaml`, we think specifying a set of modules that can be `pytest`'ed against for better flexibility. So running `bentoml test` will first serve the bento, then run the list of modules with `pytest`. In the test modules, users should be able to call the deployed bento using a client, currently being developed (https://github.com/bentoml/Kitchen/issues/42).
2. Allowing the specification of environment during `bentoml test` through the `--env` argument. Valid environments include `local` which is the same as the behavior today, `conda`, or `docker`. If `virtualenv` is specified, the `bentoml test` command will containerize the bento and run the tests in a Docker environment. The `--env` argument can also be extended beyond `bentoml test` to `bentoml serve`, and behaves the same as above.

Would you like to help with the first step of the `bentoml test` command with this PR? We are happy to share more detailed thinking over a quick chat.

aarnphm commented on bentoml/_internal/utils/buildx.py:
```diff
@@ -308,3 +308,30 @@ def run_docker_cmd(
     subprocess.check_output(list(map(str, cmds)), env=subprocess_env, cwd=cwd)
+
+
+def run_docker_container(docker_image_tag: str):
+    # TODO: allow getting arguments
+    cmd = ["docker", "run", "--rm", "-d", "-p", "3000:3000", docker_image_tag]
+    try:
+        container_id = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        container_id = container_id.decode("utf-8").strip()
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(f"Failed to run docker container: {e.output}")
+    return container_id
```
quick comment: I think we should make this as a context manager.

asafalinadsg commented:
Hi @ssheng , thanks for your replay.
Yes, I would like to help with the first step that you proposed.
I want to make sure I understand your proposal, I'll connect with you through slack.

aarnphm commented:
Hi there, any updates so far?

asafalinadsg commented:
> Hi there, any updates so far @asafalinadsg?

Hi @aarnphm 
Sorry, don't have any updates so far, didn't have the time to work on it.
I'll try to let you know in advance whenever I start work on it.

aarnphm commented:
@asafalinadsg if you don't have the bandwidth, I'm more than happy to take it over.

asafalinadsg commented:
@aarnphm Sure thing, if thats ok with you then yea I don't mind you taking over.
Thanks for the help!

parano commented:
Hi @asafalinadsg we've recently introduced the BentoML serve python API, which should make writing functional tests with Bentos a lot easier: https://github.com/bentoml/BentoML/pull/3696

Closing this PR for now as we have other plans around developer workflow that overlap with this effort.
